gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java

```diff
@@ -949,18 +964,24 @@ public void run() {
     }
 
     void removeEntry(Map regionEntryMap, RegionEntry re, LocalRegion region) {
-      if (regionEntryMap.remove(re.getKey()) == null) return;
-      if (GemFireCacheImpl.hasNewOffHeap()) {
-        // also remove reference to region buffer, if any
-        Object value = re._getValue();
-        if (value instanceof SerializedDiskBuffer) {
-          ((SerializedDiskBuffer) value).release();
+      Object removedVal = ((CustomEntryConcurrentHashMap)regionEntryMap).remove(re.getKey(),
+          re, entryRemover, null, null);
+
+      if (removedVal == null) return;
+
+      if (removedVal == re) {
+        if (GemFireCacheImpl.hasNewOffHeap()) {
+          // also remove reference to region buffer, if any
+          Object value = re._getValue();
+          if (value instanceof SerializedDiskBuffer) {
+            ((SerializedDiskBuffer)value).release();
+          }
+        }
+        // free the allocated memory
+        if (!region.reservedTable() && region.needAccounting()) {
+          NonLocalRegionEntry nre = (NonLocalRegionEntry)re;
+          region.freePoolMemory(nre.getValueSize(), nre.isForDelete());
         }
```

Comment:
The above accounting should only be done if it is a heap buffer since off-heap does it internally in its release. So set a flag after the release call above to skip this accounting for that case. This also assumes that accounting to acquire pool memory was also done and only done for heap buffers (off-heap does it during the buffer allocation itself).

Reply:
We do the accounting while storing in oldEntryMap however don't check for heap or offHeap. Will remove the accounting for off-heap at both the place.